README.md
# monit_exporter

<!-- shields.io -->
[![Contributors][contributors-shield]][contributors-url]
[![Forks][forks-shield]][forks-url]
[![Stargazers][stars-shield]][stars-url]
[![Issues][issues-shield]][issues-url]
[![MIT License][license-shield]][license-url]

## Index

* [About](#about)
  * [Features](#features)
  * [Support](#support)
  * [Dependencies](#dependencies)
* [Setup](#setup)
  * [Requirements](#requirements)
  * [Installation](#installation)
* [Usage](#usage)
* [Roadmap](#roadmap)
* [Contributing](#contributing)
* [License](#license)
* [Contact](#contact)
* [Credits](#credits)
* [Appendix](#appendix)

## About

monit_exporter periodically scrapes the monit status and provides its data via HTTP to Prometheus.

### Features

### Support

### Dependencies

This application is written in Go and has the following dependencies:
* [client_golang](github.com/prometheus/client_golang/prometheus)
* [logrus](github.com/sirupsen/logrus)
* [viper](github.com/spf13/viper)
* [charset](golang.org/x/net/html/charset)

## Setup

### Requirements

This application has the following build requirements:
* Git
* Go

### Installation

To build the application from source, simply run the following commands:
```
git clone https://github.com/liv-io/monit_exporter.git
cd monit_exporter
go build
```

## Usage

The application will load the `config.toml` file located in the same directory if present. Use the `-conf` flag to override the default configuration file name and location.

To run the application, simply execute the Go binary.

### Parameters ###

Parameter | Description | Type | Default
--- | --- | --- | ---
`listen_address` | address and port to bind | String | localhost:9388
`metrics_path` | relative path to expose metrics | String | /metrics
`ignore_ssl` | whether of not to ignore ssl errors | Boolean | false
`monit_scrape_uri` | uri to get monit status | String | http://localhost:2812/_status?format=xml&level=full
`monit_user` | user for monit basic auth, if needed | String | none
`monit_password` | password for monit status, if needed | String | none

## License

Distributed under the MIT License.

See `LICENSE` file for more information.

## Contact

Project: [monit_exporter](https://github.com/liv-io/monit_exporter)

## Credits

Acknowledgements:
* [commercetools](https://github.com/commercetools/monit_exporter)
* [delucks](https://github.com/delucks/monit_exporter)

<!-- shields.io -->
[contributors-shield]: https://img.shields.io/github/contributors/liv-io/monit_exporter.svg?style=flat
[contributors-url]: https://github.com/liv-io/monit_exporter/graphs/contributors
[forks-shield]: https://img.shields.io/github/forks/liv-io/monit_exporter.svg?style=flat
[forks-url]: https://github.com/liv-io/monit_exporter/network/members
[stars-shield]: https://img.shields.io/github/stars/liv-io/monit_exporter.svg?style=flat
[stars-url]: https://github.com/liv-io/monit_exporter/stargazers
[issues-shield]: https://img.shields.io/github/issues/liv-io/monit_exporter.svg?style=flat
[issues-url]: https://github.com/liv-io/monit_exporter/issues
[license-shield]: https://img.shields.io/github/license/liv-io/monit_exporter.svg?style=flat
[license-url]: https://github.com/liv-io/monit_exporter/blob/master/LICENSE


CHANGES
--------------------------------------------------------------------------------
INFORMATION
--------------------------------------------------------------------------------

This file lists changes made to the Ansible role. It follows semantic versioning
guidelines. The content is sorted in reverse chronological order and formatted
to allow easy grepping by scripts.

The fields are:
B: bug
C: change
E: enhancement
F: feature

--------------------------------------------------------------------------------

* Version 0.1.0 (2020-11-29)
B: Add promhttp dependency (prometheus.Handler deprecated by promhttp.Handler)
B: Improve error handling if close body is not nil
E: Return descriptive error message with HTTP response status codes
E: Remove white spaces from service types

* Version 0.0.2 (2017-10-16)
F: Change deploy key

* Version 0.0.1 (2017-09-22)
F: First working release


monit_exporter_test.go
package main

import (
	"net/http"
	"net/http/httptest"
	"testing"

	"fmt"
	"io/ioutil"
	"time"
)

const (
	monitStatus      = `<?xml version="1.0" encoding="ISO-8859-1"?><monit><server><id>acfbb9e9118e68d3754761a79d3aae16</id><incarnation>1504605214</incarnation><version>5.23.0</version><uptime>136736</uptime><poll>60</poll><startdelay>0</startdelay><localhostname>fc566edc8b68</localhostname><controlfile>/opt/monit/etc/monitrc</controlfile><httpd><address>172.17.0.2</address><port>2812</port><ssl>0</ssl></httpd></server><platform><name>Linux</name><release>4.9.27-moby</release><version>#1 SMP Thu May 11 04:01:18 UTC 2017</version><machine>x86_64</machine><cpu>4</cpu><memory>2046768</memory><swap>1048572</swap></platform><service type="5"><name>fc566edc8b68</name><collected_sec>1505209672</collected_sec><collected_usec>23215</collected_usec><status>0</status><status_hint>0</status_hint><monitor>1</monitor><monitormode>0</monitormode><onreboot>0</onreboot><pendingaction>0</pendingaction><system><load><avg01>0.00</avg01><avg05>0.00</avg05><avg15>0.00</avg15></load><cpu><user>0.1</user><system>0.1</system><wait>0.1</wait></cpu><memory><percent>6.5</percent><kilobyte>133628</kilobyte></memory><swap><percent>0.0</percent><kilobyte>0</kilobyte></swap></system></service></monit>`
	monitServiceName = `fc566edc8b68`
	monitUser        = `user`
	monitPassword    = `password`
)

func TestMonitStatus(t *testing.T) {

	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Write([]byte(monitStatus))
	})
	server := httptest.NewServer(handler)
	config := ParseConfig()
	config.monit_scrape_uri = server.URL
	e, err := NewExporter(config)
	if err != nil {
		t.Error("Unexpected error during exporter creation")
	}
	err = e.scrape()
	if err != nil {
		t.Error("Unexpected execution error:", err)
	}
}

func TestFieldsParsing(t *testing.T) {
	parsedData, err := ParseMonitStatus([]byte(monitStatus))
	if err != nil {
		t.Error("Unable to parse XML:", err)
	}
	if parsedData.MonitServices[0].Name != monitServiceName {
		t.Errorf("want Name  %s, have %s.", monitServiceName, parsedData.MonitServices[0].Name)
	}
}

func TestMonitUnavailable(t *testing.T) {
	mConfig := &Config{
		monit_scrape_uri: "http://localhost:1/status",
	}
	e, err := NewExporter(mConfig)
	if err != nil {
		t.Error("Unexpected error during exporter creation")
	}
	err = e.scrape()
	if err == nil {
		t.Error("Unexpected succsessful execution")
	}
}

func TestHttpQueryExporter(t *testing.T) {
	go main()
	time.Sleep(50 * time.Millisecond)
	address := "127.0.0.1:9388"
	resp, err := http.Get(fmt.Sprintf("http://%s/metrics", address))
	if err != nil {
		t.Fatal(err)
	}
	b, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		t.Error(err)
	}
	if err := resp.Body.Close(); err != nil {
		t.Error(err)
	}
	if want, have := http.StatusOK, resp.StatusCode; want != have {
		t.Errorf("want /metrics status code %d, have %d. Body:\n%s", want, have, b)
	}
}

func AuthHandler(w http.ResponseWriter, r *http.Request) {
	user, pass, _ := r.BasicAuth()
	if user == monitUser && pass == monitPassword {
		w.Write([]byte(monitStatus))

	} else {
		http.Error(w, "Unauthorized.", 401)
	}
}

func TestBasicAuth(t *testing.T) {
	handler := http.HandlerFunc(AuthHandler)
	server := httptest.NewServer(handler)
	config := ParseConfig()
	config.monit_scrape_uri = server.URL
	config.monit_user = monitUser
	config.monit_password = monitPassword
	e, err := NewExporter(config)
	if err != nil {
		t.Error("Unexpected error during exporter creation")
	}
	err = e.scrape()
	if err != nil {
		t.Error("Unexpected execution error:", err)
	}
}

func TestBasicAuthFail(t *testing.T) {
	handler := http.HandlerFunc(AuthHandler)
	server := httptest.NewServer(handler)
	config := ParseConfig()
	config.monit_scrape_uri = server.URL
	config.monit_user = monitUser
	config.monit_password = monitPassword + "qwe"
	e, err := NewExporter(config)
	if err != nil {
		t.Error("Unexpected error during exporter creation")
	}
	err = e.scrape()
	if err == nil {
		t.Error("Unexpected execution success:")
	}
}


monit_exporter.go
package main

import (
	"bytes"
	"crypto/tls"
	"encoding/xml"
	"errors"
	"flag"
	"fmt"
	"io/ioutil"
	"net/http"
	"sync"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	log "github.com/sirupsen/logrus"
	"github.com/spf13/viper"
	"golang.org/x/net/html/charset"
)

const (
	namespace = "monit" // Prefix for Prometheus metrics.
)

var configFile = flag.String("conf", "./config.toml", "Configuration file for exporter")

var serviceTypes = map[int]string{
	0: "filesystem",
	1: "directory",
	2: "file",
	3: "programPid",
	4: "remoteHost",
	5: "system",
	6: "fifo",
	7: "programPath",
	8: "network",
}

type monitXML struct {
	MonitServices []monitService `xml:"service"`
}

// Simplified structure of monit check.
type monitService struct {
	Type      int    `xml:"type,attr"`
	Name      string `xml:"name"`
	Status    int    `xml:"status"`
	Monitored string `xml:"monitor"`
}

// Exporter collects monit stats from the given URI and exports them using
// the prometheus metrics package.
type Exporter struct {
	config *Config
	mutex  sync.RWMutex
	client *http.Client

	up          prometheus.Gauge
	checkStatus *prometheus.GaugeVec
}

type Config struct {
	listen_address   string
	metrics_path     string
	ignore_ssl       bool
	monit_scrape_uri string
	monit_user       string
	monit_password   string
}

func FetchMonitStatus(c *Config) ([]byte, error) {
	client := &http.Client{
		Transport: &http.Transport{
			TLSClientConfig: &tls.Config{InsecureSkipVerify: c.ignore_ssl},
		},
	}

	req, err := http.NewRequest("GET", c.monit_scrape_uri, nil)
	if err != nil {
		log.Errorf("Unable to create request: %v", err)
	}

	req.SetBasicAuth(c.monit_user, c.monit_password)
	resp, err := client.Do(req)
	if err != nil {
		log.Error("Unable to fetch monit status")
		return nil, err
	}
	switch resp.StatusCode {
	case 200:
	case 401:
		return nil, errors.New("Authentication with monit failed")
	default:
		return nil, fmt.Errorf("Monit returned %s", resp.Status)
	}
	data, err := ioutil.ReadAll(resp.Body)
	defer resp.Body.Close()
	if err != nil {
		log.Fatal("Unable to read monit status")
		return nil, err
	}
	return data, nil
}

func ParseMonitStatus(data []byte) (monitXML, error) {
	var statusChunk monitXML
	reader := bytes.NewReader(data)
	decoder := xml.NewDecoder(reader)

	// Parsing status results to structure
	decoder.CharsetReader = charset.NewReaderLabel
	err := decoder.Decode(&statusChunk)
	return statusChunk, err
}

func ParseConfig() *Config {
	flag.Parse()

	v := viper.New()

	v.SetDefault("listen_address", "localhost:9388")
	v.SetDefault("metrics_path", "/metrics")
	v.SetDefault("ignore_ssl", false)
	v.SetDefault("monit_scrape_uri", "http://localhost:2812/_status?format=xml&level=full")
	v.SetDefault("monit_user", "")
	v.SetDefault("monit_password", "")
	v.SetConfigFile(*configFile)
	v.SetConfigType("toml")
	err := v.ReadInConfig() // Find and read the config file
	if err != nil {         // Handle errors reading the config file
		log.Printf("Error reading config file: %s. Using defaults.", err)
	}

	return &Config{
		listen_address:   v.GetString("listen_address"),
		metrics_path:     v.GetString("metrics_path"),
		ignore_ssl:       v.GetBool("ignore_ssl"),
		monit_scrape_uri: v.GetString("monit_scrape_uri"),
		monit_user:       v.GetString("monit_user"),
		monit_password:   v.GetString("monit_password"),
	}
}

// Returns an initialized Exporter.
func NewExporter(c *Config) (*Exporter, error) {

	return &Exporter{
		config: c,
		up: prometheus.NewGauge(prometheus.GaugeOpts{
			Namespace: namespace,
			Name:      "exporter_up",
			Help:      "Monit status availability",
		}),
		checkStatus: prometheus.NewGaugeVec(prometheus.GaugeOpts{
			Namespace: namespace,
			Name:      "exporter_service_check",
			Help:      "Monit service check info",
		},
			[]string{"check_name", "type", "monitored"},
		),
	}, nil
}

// Describe describes all the metrics ever exported by the monit exporter. It
// implements prometheus.Collector.
func (e *Exporter) Describe(ch chan<- *prometheus.Desc) {
	e.up.Describe(ch)
	e.checkStatus.Describe(ch)
}

func (e *Exporter) scrape() error {
	data, err := FetchMonitStatus(e.config)
	if err != nil {
		// set "monit_exporter_up" gauge to 0, remove previous metrics from e.checkStatus vector
		e.up.Set(0)
		e.checkStatus.Reset()
		log.Errorf("Error getting monit status: %v", err)
		return err
	} else {
		parsedData, err := ParseMonitStatus(data)
		if err != nil {
			e.up.Set(0)
			e.checkStatus.Reset()
			log.Errorf("Error parsing data from monit: %v\n%s", err, data)
		} else {
			e.up.Set(1)
			// Constructing metrics
			for _, service := range parsedData.MonitServices {
				e.checkStatus.With(prometheus.Labels{"check_name": service.Name, "type": serviceTypes[service.Type], "monitored": service.Monitored}).Set(float64(service.Status))
			}
		}
		return err
	}
}

// Collect fetches the stats from configured monit location and delivers them
// as Prometheus metrics. It implements prometheus.Collector.
func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
	e.mutex.Lock() // Protect metrics from concurrent collects.
	defer e.mutex.Unlock()
	e.checkStatus.Reset()
	e.scrape()
	e.up.Collect(ch)
	e.checkStatus.Collect(ch)
	return
}

func main() {

	config := ParseConfig()
	exporter, err := NewExporter(config)

	if err != nil {
		log.Fatal(err)
	}
	prometheus.MustRegister(exporter)

	log.Printf("Starting monit_exporter: %s", config.listen_address)
	http.Handle(config.metrics_path, promhttp.Handler())
	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		w.Write([]byte(`<html>
			<head><title>Monit Exporter</title></head>
			<body>
			<h1>Monit Exporter</h1>
			<p><a href="` + config.metrics_path + `">Metrics</a></p>
			</body>
			</html>`))
	})

	log.Fatal(http.ListenAndServe(config.listen_address, nil))
}
